xla/service/gpu/gemm_rewriter.cc

```diff
@@ -1572,7 +1572,147 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     return true;
   }
 
-  absl::Status FuseReluActivation(HloInstruction *instr,
+  absl::Status HandleSelect(HloInstruction *instr) override {
+    // Attempt to elide select and fuse gradient of ReLU into GEMM as epilogue.
+    // Rewrite the matmul as producer of bitmask into relu with Aux.
+    // 1. Check for both fwd and bwd pattern.
+    // 2. Rewrite ReLU with ReLUAux;
+    // 3. Insert DRelu as epiloguefor backward matmul.
+    HloInstruction *fwd_gemm = nullptr;
+    HloInstruction *bwd_gemm = nullptr;
+    HloInstruction *maximum = nullptr;
+    HloInstruction *compare = nullptr;
+
+    if (!Match(instr,
+               m::Select(m::Compare(&compare,
+                                    m::CustomCall(&fwd_gemm,
+                                                  {kCublasLtMatmulCallTarget}),
+                                    m::Broadcast(m::ConstantScalar(0))),
+                         m::CustomCall(&bwd_gemm, {kCublasLtMatmulCallTarget})
+                             .WithOneUser(),
+                         m::Broadcast(m::ConstantScalar(0))))) {
+      return absl::OkStatus();
+    }
+
+    if (fwd_gemm->user_count() != 2) {
+      // Require fwd_gemm only have 2 consumers: maximum and compare.
+      return absl::OkStatus();
+    }
+
+    bool valid_relu_bwd = false;
+
+    for (auto user : fwd_gemm->users()) {
+      if (user->opcode() == HloOpcode::kMaximum &&
+          Match(user,
+                m::MaximumAnyOrder(
+                    m::AnyOf<HloInstruction>(
+                        m::Slice(m::CustomCall({kCublasLtMatmulCallTarget})),
+                        m::Bitcast(m::CustomCall({kCublasLtMatmulCallTarget})),
+                        m::CustomCall({kCublasLtMatmulCallTarget})),
+                    m::Broadcast(m::ConstantScalar(0))))) {
+        maximum = user;
+        valid_relu_bwd = true;
+        break;
+      }
+    }
+    auto fwd_gemm_num_rows = fwd_gemm->shape().dimensions(0);
+    if (!valid_relu_bwd ||
+        !SupportsEpilogueFusion(fwd_gemm->shape().element_type()) ||
+        !(fwd_gemm_num_rows >= 128 && fwd_gemm_num_rows % 128 == 0)) {
+      // cublasLt requires that CUBLASLT_MATMUL_DESC_EPILOGUE_AUX_LD Must be
+      // divisible by 128 and be no less than the number of rows in the output
+      // matrix.
+      VLOG(1) << "Possible GEMM with ReLU epilogue is "
+              << "not fused into Custom Call.";
+      return absl::OkStatus();
+    }
+
+    // Rewrite Fwd Relu with ReluAux
+    TF_ASSIGN_OR_RETURN(auto gpu_config,
+                        fwd_gemm->backend_config<GpuBackendConfig>());
+    GemmBackendConfig &config = *gpu_config.mutable_gemm_backend_config();
+    if (config.epilogue() == GemmBackendConfig::DEFAULT) {
+      config.set_epilogue(GemmBackendConfig::RELU_AUX);
```

Comment:
Does this work when we fuse a matrix bias? Maybe add a test.

Reply:
Matrix bias is not supported since cublasLt's epilogue D_RELU_BGRAD only supports vector bias.

Reply:
No. hlo_verifier.cc will check the `output_to_operand_aliasing` for matrix bias C which is also D. If the output shape of forward matmul is a tuple, the check may fail.